### PR TITLE
Swift 3/Xcode 8 conversion

### PR DESCRIPTION
**Pull request by long**
Ran the conversion to Swift 3.0 against a current project using the UberRides Cocoapod. Starting point was the swift-3-dev branch.

Updated ObjectMapper spec to ~> 2.0 to have the Swift 3 version of the dependency.

Objects with type Mappable were updated to to require the named parameter and not '_' in order to match specifications in the current version of ObjectMapper

After porting the source code, the project in this repo and test files were converted in Xcode to Swift 3.0 but not built/tested

### CODE REVIEW

**Comment by long** [comment]
Project using Cocoapod builds cleanly and is able to access Uber app with the specified destination (functionality used in project app)

**Comment by jbrophy17** [comment]
Thanks for the PR! I'll go ahead and take a look at this, but before I can merge anything, could you sign the [Uber CLA](http://t.uber.com/cla)?

**Comment by kwstasna** [comment]
I got your branch and it worked for a simple app of mine. Just to open UBER with 2 locations etc. Nice work!!!

**Comment by nbgraham** [comment]
@long Thanks! This saved me a lot of headache trying to convert it myself. Now it builds successfully and runs correctly.

I had some trouble figuring out how to use @long 's version of the code in my XCode project. Here's what I did:

1. Clone @long 's [fork](https://github.com/long/rides-ios-sdk) of [uber/rides-ios-sdk](https://github.com/uber/rides-ios-sdk)  
    ```git clone https://github.com/long/rides-ios-sdk.git```
2. Switch to the [swift-3-dev branch](https://github.com/long/rides-ios-sdk/tree/swift-3-dev)  
     ```cd rides-ios-sdk```  
     ```git branch swift-3-dev ```
3. Switch to your Xcode project directory.
4. Set up Cocoapods.  
   ```pod init```
5. Add UberRides to your Podfile, **pointing to your local version** (from Step 1; Mine was in ~/Developer/rides-ios-sdk)  
   ```pod 'UberRides', :path => '~/Developer/rides-ios-sdk'```
6. Install and update your pods.  
   ```pod install```  
   ```pod update```
7. Rebuild your project!

**Comment by kwstasna** [comment]
@nbgraham Or you can do
`pod 'UberRides', :git => 'https://github.com/long/rides-ios-sdk.git', :branch => 'swift-3-dev’`

Inside your podfile and you will avoid 6 steps! :dancer:

**Comment by nbgraham** [comment]
@kwstasna Thanks, that's a lot better.

I got a lot of build errors the first time, until I used XCode 8 [migration tool](https://swift.org/migration-guide/)  *Edit -> Convert -> To Current Swift Syntax…*

Then it said no update necessary, and I rebuilt to get it working.

Why doesn't it work straight away?

**Comment by kwstasna** [comment]
@nbgraham I really don't know. I can't reproduce because as soon as I installed the pod as I told you above, I got no errors from Uber Rides!

**Comment by jbrophy17** [comment]
@long Again thanks for this PR, but I need you to sign the [Uber CLA](http://t.uber.com/cla) before I can merge this in

File: source/UberRides/OAuth/KeychainWrapper.swift [hunk 1]
```diff
@@ -24,25 +24,25 @@
...
-    private let MatchLimit = kSecMatchLimit as String
+    fileprivate let Class = kSecClass as String
+    fileprivate let AttrAccount = kSecAttrAccount as String
+    fileprivate let AttrService = kSecAttrService as String
+    fileprivate let AttrAccessGroup = kSecAttrAccessGroup as String
+    fileprivate let AttrGeneric = kSecAttrGeneric as String
+    fileprivate let AttrAccessible = kSecAttrAccessible as String
+    fileprivate let ReturnData = kSecReturnData as String
+    fileprivate let ValueData = kSecValueData as String
+    fileprivate let MatchLimit = kSecMatchLimit as String
```
Comment (jbrophy17) [inline source/UberRides/OAuth/KeychainWrapper.swift]: These can all be `private`

File: source/UberRides/OAuth/KeychainWrapper.swift [hunk 2]
```diff
@@ -117,17 +117,17 @@ class KeychainWrapper: NSObject {
      
      - returns: dictionary of base attributes for keychain query.
      */
-    private func getKeychainItemData(key: String) -> [String: AnyObject] {
+    fileprivate func getKeychainItemData(_ key: String) -> [String: AnyObject] {
```
Comment (jbrophy17) [inline source/UberRides/OAuth/KeychainWrapper.swift]: can be `private`

File: source/UberRides/OAuth/LoginManager.swift [hunk 3]
```diff
@@ -205,39 +205,39 @@
     
     // Mark: Private Interface
     
-    private func handleLoginCanceled() {
+    fileprivate func handleLoginCanceled() {
```
Comment (jbrophy17) [inline source/UberRides/OAuth/LoginManager.swift]: can be `private`

File: source/UberRides/OAuth/LoginManager.swift [hunk 4]
```diff
@@ -205,39 +205,39 @@
...
         loggingIn = false
-        authenticator?.loginCompletion?(accessToken: nil, error: RidesAuthenticationErrorFactory.errorForType(ridesAuthenticationErrorType: .UserCancelled))
+        authenticator?.loginCompletion?(nil, RidesAuthenticationErrorFactory.errorForType(ridesAuthenticationErrorType: .userCancelled))
         authenticator = nil
     }
     
-    private func loginCompletion(loginType: LoginType, presentingViewController: UIViewController?, completion: ((accessToken: AccessToken?, error: NSError?) -> Void)?) -> ((accessToken: AccessToken?, error: NSError?) -> Void)? {
-        var loginCompletion: ((accessToken: AccessToken?, error: NSError?) -> Void)?
+    fileprivate func loginCompletion(_ loginType: LoginType, presentingViewController: UIViewController?, completion: ((_ accessToken: AccessToken?, _ error: NSError?) -> Void)?) -> ((_ accessToken: AccessToken?, _ error: NSError?) -> Void)? {
```
Comment (jbrophy17) [inline source/UberRides/OAuth/LoginManager.swift]: can be `private`

File: source/UberRides/OAuth/LoginManager.swift [hunk 5]
```diff
@@ -246,24 +246,24 @@
         return loginCompletion
     }
     
-    private func handleNativeFallback(error: NSError?, presentingViewController: UIViewController?, completion: ((accessToken: AccessToken?, error: NSError?) -> Void)?) {
+    fileprivate func handleNativeFallback(_ error: NSError?, presentingViewController: UIViewController?, completion: ((_ accessToken: AccessToken?, _ error: NSError?) -> Void)?) {
```
Comment (jbrophy17) [inline source/UberRides/OAuth/LoginManager.swift]: can be `private`

File: source/UberRides/RequestDeeplink.swift [hunk 6]
```diff
@@ -29,10 +29,10 @@ import UIKit
 /**
  *  Builds and executes a deeplink to the native Uber app to request a ride.
  */
-@objc(UBSDKRequestDeeplink) public class RequestDeeplink: BaseDeeplink {
+@objc(UBSDKRequestDeeplink) open class RequestDeeplink: BaseDeeplink {
     
-    private let parameters: RideParameters
-    private let clientID: String
+    fileprivate let parameters: RideParameters
+    fileprivate let clientID: String
```
Comment (jbrophy17) [inline source/UberRides/RequestDeeplink.swift]: these can both be `private`

File: source/UberRides/RequestURLBuilder.swift [hunk 7]
```diff
@@ -26,7 +26,7 @@ import CoreLocation
 
 class RequestURLBuilder {
     
-    private enum LocationType: String {
+    fileprivate enum LocationType: String {
```
Comment (jbrophy17) [inline source/UberRides/RequestURLBuilder.swift]: can be `private`

File: source/UberRides/RequestURLBuilder.swift [hunk 8]
```diff
@@ -43,66 +43,66 @@ class RequestURLBuilder {
     static let deeplinkScheme = "uber"
     static let userAgentKey = "user-agent"
     
-    private let clientID: String
-    private var rideParameters: RideParameters?
+    fileprivate let clientID: String
+    fileprivate var rideParameters: RideParameters?
```
Comment (jbrophy17) [inline source/UberRides/RequestURLBuilder.swift]: both can be `private`

File: source/UberRides/RequestURLBuilder.swift [hunk 9]
```diff
@@ -43,66 +43,66 @@ class RequestURLBuilder {
...
         
         urlComponents.queryItems = queryItems
         
-        return urlComponents.URL
+        return urlComponents.url
     }
     
-    private func addLocation(locationType: LocationType, location: CLLocation, nickname: String?, address: String?) -> [NSURLQueryItem] {
-        var queryItems = [NSURLQueryItem]()
+    fileprivate func addLocation(_ locationType: LocationType, location: CLLocation, nickname: String?, address: String?) -> [URLQueryItem] {
```
Comment (jbrophy17) [inline source/UberRides/RequestURLBuilder.swift]: can be `private`

File: source/UberRides/RideRequestButton.swift [hunk 10]
```diff
@@ -34,37 +34,37 @@ import CoreLocation
...
-    public var client: RidesClient?
+    open var client: RidesClient?
     
     static let sourceString = "button"
     
     var metadata: ButtonMetadata = ButtonMetadata()
     var uberMetadataLabel: UILabel = UILabel()
     
-    private let opticalCorrection: CGFloat = 1.0
+    fileprivate let opticalCorrection: CGFloat = 1.0
```
Comment (jbrophy17) [inline source/UberRides/RideRequestButton.swift]: can be `private`

File: source/UberRides/RideRequestButton.swift [hunk 11]
```diff
@@ -307,50 +307,50 @@ import CoreLocation
      - parameter subtitle: The subtitle of the label. (ex. "$6-8 for uberX")
      - parameter surge:    Whether the price estimate should include a surge image. Default false.
      */
-    private func setMultilineAttributedString(title: String, subtitle: String = "", surge: Bool = false) {
-        let metadataFont = UIFont(name: "HelveticaNeue-Regular", size: 12) ?? UIFont.systemFontOfSize(12)
+    fileprivate func setMultilineAttributedString(_ title: String, subtitle: String = "", surge: Bool = false) {
```
Comment (jbrophy17) [inline source/UberRides/RideRequestButton.swift]: can be `private`

File: source/UberRides/RideRequestButton.swift [hunk 12]
```diff
@@ -456,18 +456,18 @@ import CoreLocation
     }
     
     // get image from media directory
-    private func getImage(name: String) -> UIImage {
-        let bundle = NSBundle(forClass: RideRequestButton.self)
-        let image = UIImage(named: name, inBundle: bundle, compatibleWithTraitCollection: nil)
+    fileprivate func getImage(_ name: String) -> UIImage {
```
Comment (jbrophy17) [inline source/UberRides/RideRequestButton.swift]: can be `private`

File: source/UberRides/RideRequestButton.swift [hunk 13]
```diff
@@ -481,7 +481,7 @@ struct ButtonMetadata {
     var dropoffLatitude: Double?
     var dropoffLongitude: Double?
     var timeEstimate: TimeEstimate?
-    private var priceEstimateList: [PriceEstimate]?
+    fileprivate var priceEstimateList: [PriceEstimate]?
```
Comment (jbrophy17) [inline source/UberRides/RideRequestButton.swift]: priceEstimateList

File: source/UberRides/RideRequestView.swift [hunk 14]
```diff
@@ -156,100 +156,100 @@ import CoreLocation
...
-            webView.loadRequest(NSURLRequest(URL: url))
+        if let url = URL(string: "about:blank") {
+            webView.load(URLRequest(url: url))
         }
     }
     
     // MARK: Private
     
-    private func initialSetup() {
+    fileprivate func initialSetup() {
```
Comment (jbrophy17) [inline source/UberRides/RideRequestView.swift]: can be `private`

File: source/UberRides/RideRequestView.swift [hunk 15]
```diff
@@ -156,100 +156,100 @@ import CoreLocation
...
-        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(keyboardWillAppear(_:)), name: UIKeyboardWillShowNotification, object: nil)
-        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(keyboardDidAppear(_:)), name: UIKeyboardDidShowNotification, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillAppear(_:)), name: NSNotification.Name.UIKeyboardWillShow, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(keyboardDidAppear(_:)), name: NSNotification.Name.UIKeyboardDidShow, object: nil)
         
         setupWebView()
     }
     
-    private func setupWebView() {
+    fileprivate func setupWebView() {
```
Comment (jbrophy17) [inline source/UberRides/RideRequestView.swift]: can be `private`

File: source/UberRides/RideRequestViewController.swift [hunk 16]
```diff
@@ -59,10 +59,10 @@ import MapKit
...
     static let sourceString = "ride_request_widget"
     
-    private var accessTokenWasUnauthorizedOnPreviousAttempt = false
-    private var accessTokenIdentifier: String
-    private var keychainAccessGroup: String
-    private var loginCompletion: ((accessToken: AccessToken?, error: NSError?) -> Void)?
+    fileprivate var accessTokenWasUnauthorizedOnPreviousAttempt = false
+    fileprivate var accessTokenIdentifier: String
+    fileprivate var keychainAccessGroup: String
+    fileprivate var loginCompletion: ((_ accessToken: AccessToken?, _ error: NSError?) -> Void)?
```
Comment (jbrophy17) [inline source/UberRides/RideRequestViewController.swift]: `loginCompletion` can be `private`

**Comment by jbrophy17** [comment]
Also, @long you need you to sign the [Uber CLA](http://t.uber.com/cla) before I can merge this

**Comment by long** [comment]
Sorry, I hadn't logged into github directly for a bit and hadn't received email notifications.

I've just signed the Uber CLA

**Comment by jbrophy17** [comment]
Hey @long, have you had a chance to look at my comments on your PR? Would love to get this merged!

**Comment by vinayjn** [comment]
@jbrophy17 when are you planning to release the Swift 3 SDK, with the release of Xcode 8.3, I cannot build the SDK with Swift 2.3.

**Comment by vinayjn** [comment]
Really sad to see inactive development for this SDK. Swift 4 is released and the SDK is still not updated for Swift 3.

**Comment by arogal** [comment]
Hi Vinay,

Thank you for your patience. We are indeed picking back up on this in a few
weeks. We'll be posting a plan for Swift 3/4 so everyone can keep informed.

Thanks.

**Comment by edjiang** [comment]
Closing since it looks like the bulk of the migration was taken care of in 892a779a38dfefaf364f9b79ad9142e7a45ffe45

Sorry this took so long, and thank you @long for all of your help here :) 

cc @vinayjn since you were asking for updates :)

See #167 for installing the Swift 3 package via your package manager.

**Pull request closed without merging**